tensorflow_addons/text/tests/crf_test.py

```diff
@@ -501,6 +502,10 @@ def test_tf_function():
     tf.__version__[:3] == "2.4",
     reason="CRF Decode doesn't work in TF2.4, the issue was fixed in TF core, but didn't make the release",
 )
+@pytest.mark.skipif(
+    tf.__version__[:3] == "2.5",
+    reason="CRF decoding models have serialization issues in TF >=2.5 . Please see isse #",
+)
```

Comment:
Need to create an issue for this log which fails on the loading of serialized model:
Looks to be related to https://github.com/tensorflow/tensorflow/pull/45534
```
=========================================================================================== FAILURES ============================================================================================
________________________________________________________________________________ test_crf_decode_save_load[cpu] _________________________________________________________________________________

tmpdir = local('/tmp/pytest-of-root/pytest-3/test_crf_decode_save_load_cpu_0')

    @pytest.mark.skipif(
        tf.__version__[:3] == "2.4",
        reason="CRF Decode doesn't work in TF2.4, the issue was fixed in TF core, but didn't make the release",
    )
    def test_crf_decode_save_load(tmpdir):
        tf.keras.backend.clear_session()
        input_tensor = tf.keras.Input(shape=(10, 3), dtype=tf.float32, name="input_tensor")
        seq_len = tf.keras.Input(shape=(), dtype=tf.int32, name="seq_len")
        transition = tf.constant([[1, 1, 0], [0, 1, 1], [1, 0, 1]], dtype=tf.float32)

        output = tf.multiply(input_tensor, tf.constant(1.0))
        decoded, _ = text.crf_decode(input_tensor, transition, seq_len)

        model = tf.keras.Model(
            inputs=[input_tensor, seq_len], outputs=[output, decoded], name="example_model"
        )
        model.compile(optimizer="Adam")

        x_data = {
            "input_tensor": np.random.random_sample((5, 10, 3)).astype(dtype=np.float32),
            "seq_len": np.array([10] * 5, dtype=np.int32),
        }
        y_data = {"tf.math.multiply": np.random.randint(0, 3, (5, 10))}

        model.fit(x_data, y_data)
        model.predict(
            {
                "input_tensor": tf.expand_dims(x_data["input_tensor"][0], 0),
                "seq_len": np.array([10]),
            }
        )

        temp_dir = str(tmpdir.mkdir("model"))
        tf.saved_model.save(model, temp_dir)

        tf.keras.backend.clear_session()
>       model = tf.keras.models.load_model(
            temp_dir,
            custom_objects={"CrfDecodeForwardRnnCell": text.crf.CrfDecodeForwardRnnCell},
        )

tensorflow_addons/text/tests/crf_test.py:536:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/saving/save.py:206: in load_model
    return saved_model_load.load(filepath, compile, options)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/load.py:155: in load
    keras_loader.finalize_objects()
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/load.py:626: in finalize_objects
    self._reconstruct_all_models()
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/load.py:645: in _reconstruct_all_models
    self._reconstruct_model(model_id, model, layers)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/load.py:691: in _reconstruct_model
    created_layers) = functional_lib.reconstruct_from_config(
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/engine/functional.py:1289: in reconstruct_from_config
    process_node(layer, node_data)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/engine/functional.py:1237: in process_node
    output_tensors = layer(input_tensors, **kwargs)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py:969: in __call__
    return self._functional_construction_call(inputs, args, kwargs,
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py:1107: in _functional_construction_call
    outputs = self._keras_tensor_symbolic_call(
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py:840: in _keras_tensor_symbolic_call
    return self._infer_output_signature(inputs, args, kwargs, input_masks)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py:880: in _infer_output_signature
    outputs = call_fn(inputs, *args, **kwargs)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py:1363: in _call_wrapper
    return self._call_wrapper(*args, **kwargs)
/usr/local/lib/python3.8/site-packages/tensorflow/python/keras/layers/core.py:1395: in _call_wrapper
    result = self.function(*args, **kwargs)
/usr/local/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py:206: in wrapper
    return target(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

fn = '_scan_fn', elems = <tf.Tensor 'Placeholder:0' shape=(9, None, 3) dtype=int32>, initializer = <tf.Tensor 'Placeholder_1:0' shape=(None, 1) dtype=int32>, parallel_iterations = 10
back_prop = True, swap_memory = False, infer_shape = True, reverse = False, name = None

    @tf_export(v1=["scan"])
    @dispatch.add_dispatch_support
    def scan(fn,
             elems,
             initializer=None,
             parallel_iterations=10,
             back_prop=True,
             swap_memory=False,
             infer_shape=True,
             reverse=False,
             name=None):
      """scan on the list of tensors unpacked from `elems` on dimension 0.

      See also `tf.map_fn`.

      The simplest version of `scan` repeatedly applies the callable `fn` to a
      sequence of elements from first to last. The elements are made of the tensors
      unpacked from `elems` on dimension 0. The callable fn takes two tensors as
      arguments. The first argument is the accumulated value computed from the
      preceding invocation of fn, and the second is the value at the current
      position of `elems`. If `initializer` is None, `elems` must contain at least
      one element, and its first element is used as the initializer.

      Suppose that `elems` is unpacked into `values`, a list of tensors. The shape
      of the result tensor is `[len(values)] + fn(initializer, values[0]).shape`.
      If reverse=True, it's fn(initializer, values[-1]).shape.

      This method also allows multi-arity `elems` and accumulator.  If `elems`
      is a (possibly nested) list or tuple of tensors, then each of these tensors
      must have a matching first (unpack) dimension.  The second argument of
      `fn` must match the structure of `elems`.

      If no `initializer` is provided, the output structure and dtypes of `fn`
      are assumed to be the same as its input; and in this case, the first
      argument of `fn` must match the structure of `elems`.

      If an `initializer` is provided, then the output of `fn` must have the same
      structure as `initializer`; and the first argument of `fn` must match
      this structure.

      For example, if `elems` is `(t1, [t2, t3])` and `initializer` is
      `[i1, i2]` then an appropriate signature for `fn` in `python2` is:
      `fn = lambda (acc_p1, acc_p2), (t1, [t2, t3]):` and `fn` must return a list,
      `[acc_n1, acc_n2]`.  An alternative correct signature for `fn`, and the
       one that works in `python3`, is:
      `fn = lambda a, t:`, where `a` and `t` correspond to the input tuples.

      Args:
        fn: The callable to be performed.  It accepts two arguments.  The first will
          have the same structure as `initializer` if one is provided, otherwise it
          will have the same structure as `elems`.  The second will have the same
          (possibly nested) structure as `elems`.  Its output must have the same
          structure as `initializer` if one is provided, otherwise it must have the
          same structure as `elems`.
        elems: A tensor or (possibly nested) sequence of tensors, each of which will
          be unpacked along their first dimension.  The nested sequence of the
          resulting slices will be the first argument to `fn`.
        initializer: (optional) A tensor or (possibly nested) sequence of tensors,
          initial value for the accumulator, and the expected output type of `fn`.
        parallel_iterations: (optional) The number of iterations allowed to run in
          parallel.
        back_prop: (optional) True enables support for back propagation.
        swap_memory: (optional) True enables GPU-CPU memory swapping.
        infer_shape: (optional) False disables tests for consistent output shapes.
        reverse: (optional) True scans the tensor last to first (instead of first to
          last).
        name: (optional) Name prefix for the returned tensors.

      Returns:
        A tensor or (possibly nested) sequence of tensors.  Each tensor packs the
        results of applying `fn` to tensors unpacked from `elems` along the first
        dimension, and the previous accumulator value(s), from first to last (or
        last to first, if `reverse=True`).

      Raises:
        TypeError: if `fn` is not callable or the structure of the output of
          `fn` and `initializer` do not match.
        ValueError: if the lengths of the output of `fn` and `initializer`
          do not match.

      Examples:
        ```python
        elems = np.array([1, 2, 3, 4, 5, 6])
        sum = scan(lambda a, x: a + x, elems)
        # sum == [1, 3, 6, 10, 15, 21]
        sum = scan(lambda a, x: a + x, elems, reverse=True)
        # sum == [21, 20, 18, 15, 11, 6]
        ```

        ```python
        elems = np.array([1, 2, 3, 4, 5, 6])
        initializer = np.array(0)
        sum_one = scan(
            lambda a, x: x[0] - x[1] + a, (elems + 1, elems), initializer)
        # sum_one == [1, 2, 3, 4, 5, 6]
        ```

        ```python
        elems = np.array([1, 0, 0, 0, 0, 0])
        initializer = (np.array(0), np.array(1))
        fibonaccis = scan(lambda a, _: (a[1], a[0] + a[1]), elems, initializer)
        # fibonaccis == ([1, 1, 2, 3, 5, 8], [1, 2, 3, 5, 8, 13])
        ```
      """
      if not callable(fn):
>       raise TypeError("fn must be callable.")
E       TypeError: fn must be callable.

/usr/local/lib/python3.8/site-packages/tensorflow/python/ops/functional_ops.py:544: TypeError
```